Fix code coverage reporting

Today I noticed that our code coverage reporting has been broken for ages for both coveralls.io and codecov.io. I didn't have time to look into this, but I guess something's wrong with our TravisCI config and we're not submitting any data there. 

On a related note - probably we should simply remove coveralls.io and use only codecov.io, as it seems like a nicer service. I don't even recall why were submitting data to both of them to begin with.

This looks to me like a variation on #80, where in this case cloverage is throwing an exception:
```
Caused by: java.lang.ClassNotFoundException: com.sun.javadoc.ClassDoc
	at java.net.URLClassLoader.findClass(URLClassLoader.java:381)
	at clojure.lang.DynamicClassLoader.findClass(DynamicClassLoader.java:69)
	at java.lang.ClassLoader.loadClass(ClassLoader.java:424)
	at clojure.lang.DynamicClassLoader.loadClass(DynamicClassLoader.java:77)
	at java.lang.ClassLoader.loadClass(ClassLoader.java:357)
	at java.lang.Class.forName0(Native Method)
	at java.lang.Class.forName(Class.java:348)
	at clojure.lang.RT.classForName(RT.java:2154)
	at clojure.lang.RT.classForNameNonLoading(RT.java:2167)
	at cider.nrepl.middleware.util.java.parser$eval8009$loading__5340__auto____8010.invoke(cider/nrepl/middleware/util/java/parser.clj:1)
	at cider.nrepl.middleware.util.java.parser$eval8009.invoke(cider/nrepl/middleware/util/java/parser.clj:1)
	at clojure.lang.Compiler.eval(Compiler.java:6782)
	at clojure.lang.Compiler.eval(Compiler.java:6771)
	at clojure.lang.Compiler.eval(Compiler.java:6772)
	at clojure.lang.Compiler.eval(Compiler.java:6745)
	at clojure.core$eval.invoke(core.clj:3081)
	at cloverage.instrument$instrument$fn__2088$fn__2089.invoke(instrument.clj:428)
	at cloverage.instrument$instrument$fn__2088.invoke(instrument.clj:426)
	... 20 more
```

I've investigated further. 

In a nutshell, `cloverage` instrumentation is failing on certain namespaces, and/or causing certain tests to fail. Some of these issues are obscuring the others, so I spent the time to selectively exclude namespaces from `cloverage` instrumentation until I could get tests to run.

Note: I haven't found a convenient way to run `cloverage` in a repl and/or step through in a debugger, so I've cloned it and added some instrumentation with `dgrnbrg/spyscope`.  In what follows, `CLOVERAGE_VERSION=1.0.11-SNAPSHOT` is my local build.

1. Running `cloverage` exactly as in `.travis.yml` throws `ClassNotFoundException` as shown in my earlier comment. Presumably this is because `tools.jar` is not in the classpath when `cloverage` attempts to instrument `cider.nrepl.middleware.util.java.parser`, and so we're going to have to arrange things so that jar is on the classpath first.

2. I manually excluded the namespace `cider.nrepl.middleware.util.java.parser` from instrumentation and that revealed a second problem:

```
$ CLOVERAGE_VERSION=1.0.11-SNAPSHOT lein with-profile +1.7,+cloverage,+test-clj cloverage -e ".*java.parser"
<... output indicating namespaces being loaded and instrumented ...>
Exception in thread "main" java.lang.Exception: Couldn't eval form (cloverage.instrument/wrapm
cloverage.coverage/track-coverage 1 ^{:line 1, :column 1, :end-line 3, :end-column 45, :file
"cider_nrepl/plugin.clj"} (^{:line 1, :column 2, :end-line 1, :end-column 4} ns ^{:line 1, :column 5, 
:end-line 1, :end-column 23} cider-nrepl.plugin ^{:line 2, :column 3, :end-line 3, :end-column 44} 
(:require ^{:line 2, :column 13, :end-line 2, :end-column 37} [^{:line 2, :column 14, :end-line 2, :end-
column 29} clojure.java.io :as ^{:line 2, :column 34, :end-line 2, :end-column 36} io] ^{:line 3, 
:column 13, :end-line 3, :end-column 43} [^{:line 3, :column 14, :end-line 3, :end-column 33} 
leiningen.core.main :as ^{:line 3, :column 38, :end-line 3, :end-column 42} lein])))
<...>
Caused by: java.lang.IllegalArgumentException: No implementation of method: :make-reader of 
protocol: #'clojure.java.io/IOFactory found for class: nil, compiling:(classpath.clj:407:35)
```
I haven't tried to understand what that exception is about.

3. Pushing ahead, I also excluded the namespace `cider-nrepl.plugin` from instrumentation such that `cloverage` actually gets to running the tests.  Third issue, there's a stack overflow in `cider.nrepl.middleware.util.instrument-test`.

```
CLOVERAGE_VERSION=1.0.11-SNAPSHOT lein with-profile +1.7,+cloverage,+test-clj cloverage -e ".*java.parser" -e "cider-nrepl.plugin"
<... output about loading / instrumenting the namespaces ...>
Instrumented namespaces.

Testing cider.nrepl.middleware.classpath-test

Testing cider.nrepl.middleware.undef-test

Testing cider.nrepl.middleware.util.instrument-test

ERROR in (namespace-embedded-in-code-test) (Util.java:173)
Uncaught exception, not in assertion.
expected: nil
  actual: java.lang.StackOverflowError: null
 at clojure.lang.Util.hasheq (Util.java:173)
    clojure.lang.PersistentHashMap.hash (PersistentHashMap.java:120)
    clojure.lang.PersistentHashMap.entryAt (PersistentHashMap.java:132)
    clojure.lang.Var.getThreadBinding (Var.java:355)
    clojure.lang.Var.deref (Var.java:195)
    clojure.lang.Var.get (Var.java:191)
    clojure.core$pr.invoke (core.clj:3560)
    clojure.lang.AFn.applyToHelper (AFn.java:154)
    clojure.lang.RestFn.applyTo (RestFn.java:132)
    clojure.core$apply.invoke (core.clj:630)
    clojure.core$pr_str.doInvoke (core.clj:4563)
    clojure.lang.RestFn.invoke (RestFn.java:408)
    cider.nrepl.middleware.util.instrument$fn__6784$fn__6790.invoke (cider/nrepl/middleware/util/instrument.clj:233)
    clojure.core$map$fn__4553.invoke (core.clj:2624)
    clojure.lang.LazySeq.sval (LazySeq.java:40)
    clojure.lang.LazySeq.seq (LazySeq.java:49)
    clojure.lang.RT.seq (RT.java:507)
    clojure.core/seq (core.clj:137)
    clojure.core.protocols$seq_reduce.invoke (protocols.clj:30)
    clojure.core.protocols/fn (protocols.clj:101)
    clojure.core.protocols$fn__6452$G__6447__6465.invoke (protocols.clj:13)
    clojure.core$reduce.invoke (core.clj:6519)
    clojure.core$into.invoke (core.clj:6600)
    cider.nrepl.middleware.util.instrument/fn (cider/nrepl/middleware/util/instrument.clj:232)
    cider.nrepl.middleware.util.instrument/fn (cider/nrepl/middleware/util/instrument.clj:224)
    cider.nrepl.middleware.util.instrument/fn (cider/nrepl/middleware/util/instrument.clj:293)
    cider.nrepl.middleware.util.instrument$fn__6816$fn__6817.invoke (cider/nrepl/middleware/util/instrument.clj:305)
    cider.nrepl.middleware.util.instrument/fn (cider/nrepl/middleware/util/instrument.clj:243)
    cider.nrepl.middleware.util.instrument/fn (cider/nrepl/middleware/util/instrument.clj:224)
    cider.nrepl.middleware.util.instrument/fn (cider/nrepl/middleware/util/instrument.clj:293)
    cider.nrepl.middleware.util.instrument$fn__6816$fn__6817.invoke (cider/nrepl/middleware/util/instrument.clj:305)
    cider.nrepl.middleware.util.instrument/fn (cider/nrepl/middleware/util/instrument.clj:243)
    cider.nrepl.middleware.util.instrument/fn (cider/nrepl/middleware/util/instrument.clj:224)
    cider.nrepl.middleware.util.instrument/fn (cider/nrepl/middleware/util/instrument.clj:293)
    cider.nrepl.middleware.util.instrument$fn__6816$fn__6817.invoke (cider/nrepl/middleware/util/instrument.clj:305)
<... repeats ...>
```

Caused by some sort of unbounded recursion in `strip-instrumentation-meta`.  Intuitively I'm not surprised if cloverage instrumentation collides with the instrumentation in this middleware, but I didn't try to understand further.

4. I excluded `cider.nrepl.middleware.util.instrument` also, to see what else:

```
CLOVERAGE_VERSION=1.0.11-SNAPSHOT lein with-profile +1.7,+cloverage,+test-clj cloverage -e ".*java.parser" -e "cider-nrepl.plugin" -e ".*util.instrument"
<... test output omitted ...>
Ran 180 tests containing 1091 assertions.
39 failures, 3 errors.
Ran tests.
Writing HTML report to: /Users/neilo/Documents/code/cider-nrepl/target/coverage/index.html

|--------------------------------------------------+---------+---------|
|                                        Namespace | % Forms | % Lines |
|--------------------------------------------------+---------+---------|
|                                      cider.nrepl |   96.19 |   99.59 |
|                   cider.nrepl.middleware.apropos |   97.49 |  100.00 |
|                 cider.nrepl.middleware.classpath |   90.77 |  100.00 |
|                  cider.nrepl.middleware.complete |   84.87 |   92.31 |
|                     cider.nrepl.middleware.debug |   82.63 |   82.62 |
| cider.nrepl.middleware.debug-integration-test.fn |   71.43 |   85.71 |
|                 cider.nrepl.middleware.enlighten |    2.33 |   19.05 |
|                    cider.nrepl.middleware.format |   94.59 |  100.00 |
|                      cider.nrepl.middleware.info |   89.97 |   90.86 |
|                   cider.nrepl.middleware.inspect |   91.39 |   92.98 |
|               cider.nrepl.middleware.macroexpand |   61.34 |   62.40 |
|                        cider.nrepl.middleware.ns |   69.73 |   72.60 |
|                       cider.nrepl.middleware.out |   38.19 |   43.40 |
|                    cider.nrepl.middleware.pprint |   99.07 |  100.00 |
|                   cider.nrepl.middleware.refresh |   94.77 |   98.06 |
|                  cider.nrepl.middleware.resource |   95.27 |   97.44 |
|                      cider.nrepl.middleware.spec |   42.65 |   51.43 |
|                cider.nrepl.middleware.stacktrace |   66.60 |   78.16 |
|                      cider.nrepl.middleware.test |   19.94 |   21.29 |
|                     cider.nrepl.middleware.trace |   94.87 |  100.00 |
|               cider.nrepl.middleware.track-state |   96.88 |   95.96 |
|                     cider.nrepl.middleware.undef |   90.91 |  100.00 |
|          cider.nrepl.middleware.util.classloader |  100.00 |  100.00 |
|                 cider.nrepl.middleware.util.cljs |   51.61 |   67.31 |
|       cider.nrepl.middleware.util.error-handling |   99.23 |  100.00 |
|                 cider.nrepl.middleware.util.java |   95.74 |   97.14 |
|                 cider.nrepl.middleware.util.meta |   82.85 |   88.82 |
|                 cider.nrepl.middleware.util.misc |   98.90 |   98.33 |
|            cider.nrepl.middleware.util.namespace |   92.98 |   95.38 |
|                cider.nrepl.middleware.util.nrepl |    5.13 |   25.00 |
|                 cider.nrepl.middleware.util.spec |   69.23 |  100.00 |
|                   cider.nrepl.middleware.version |   89.66 |   87.50 |
|                         cider.nrepl.print-method |   89.55 |   94.87 |
|                              cider.nrepl.version |  100.00 |  100.00 |
|                                      cider.tasks |   20.05 |   22.22 |
|--------------------------------------------------+---------+---------|
|                                        ALL FILES |   73.27 |   81.69 |
|--------------------------------------------------+---------+---------|
Error encountered performing task 'cloverage' with profile(s): 'base,system,user,provided,dev,1.7,cloverage,test-clj'
Suppressed exit
```

And finally we've got tests.  39 test failures and 3 errors occurring in "debug-integration-test". The same tests run successfully without `cloverage` in the mix. 


So in summary, the four issues are:

1. Exception thrown when `cloverage` instruments `cider.nrepl.middleware.util.java.parser`, because `tools.jar` not on the classpath.

2. Exception thrown when `cloverage` instruments `cider-nrepl.plugin`. Reasons unknown.

3. Stack overflow when `cloverage` executes `cider.nrepl.middleware.util.instrument-test`, provided `cider.nrepl.middleware.util.instrument` is instrumented. Some sort of infinite recursion in `strip-instrumentation-meta`.

4. Multiple test errors and failures occuring in `cider.nrepl.middleware.debug-integration-test` that occur only when tests run in `cloverage`. Reasons unknown.

--- a/.gitignore
+++ b/.gitignore
@@ -12,3 +12,4 @@ pom.xml.asc
 .lein-plugins
 .lein-repl-history
 .nrepl-port
+.source-deps

--- a/.travis.yml
+++ b/.travis.yml
@@ -5,69 +5,52 @@ cache:
   directories:
     - $HOME/.m2
 script:
-  - lein source-deps
-  - |
-    if [ ! -z "$TEST_CMD" ]; then
-      $TEST_CMD;
-    else
-      lein with-profile +$VERSION,+plugin.mranderson/config,+$TEST test;
-    fi
+  - make $TARGET
 env:
-  - VERSION="1.7" TEST="test-clj"
-  - VERSION="1.7" TEST="test-cljs"
-  - VERSION="1.8" TEST="test-clj"
-  - VERSION="1.8" TEST="test-cljs"
-  - VERSION="1.9" TEST="test-clj"
-  - VERSION="1.9" TEST="test-cljs"
+  - VERSION=1.7 TARGET=test-clj
+  - VERSION=1.7 TARGET=test-cljs
+  - VERSION=1.8 TARGET=test-clj
+  - VERSION=1.8 TARGET=test-cljs
+  - VERSION=1.9 TARGET=test-clj
+  - VERSION=1.9 TARGET=test-cljs
 jdk:
   - oraclejdk7
   - oraclejdk8
+  - oraclejdk9
 matrix:
   include:
     # Test OpenJDK against latest Clojure stable
-    - env:
-        - VERSION="1.9"
-        - TEST="test-clj"
+    - env: VERSION=1.9 TARGET=test-clj
       jdk: openjdk8
-    - env:
-        - VERSION="1.9"
-        - TEST="test-cljs"
+    - env: VERSION=1.9 TARGET=test-cljs
       jdk: openjdk8
 
     # Test Clojure master against a single JDK
-    - env:
-        - VERSION="master"
-        - TEST="test-clj"
+    - env: VERSION=master TARGET=test-clj
       jdk: oraclejdk8
-    - env:
-        - VERSION="master"
-        - TEST="test-cljs"
+    - env: VERSION=master TARGET=test-cljs
       jdk: oraclejdk8
 
-    # eastwood
-    - env: TEST_CMD="lein with-profile +1.9,+test-clj,+test-cljs,+eastwood eastwood"
+    # Eastwood linter
+    - env: VERSION=1.9 TARGET=eastwood
       jdk: oraclejdk8
 
-    # cljfmt
-    - env: TEST_CMD="lein with-profile +1.9,+test-clj,+test-cljs,+cljfmt cljfmt check"
+    # Check cljfmt
+    - env: VERSION=1.9 TARGET=cljfmt
       jdk: oraclejdk8
 
-    # codecov
-    - env: TEST_CMD="lein with-profile +1.9,+test-clj,+test-cljs,+cloverage cloverage --codecov"
+    # Coverage analysis
+    - env: VERSION=1.9 TARGET=cloverage
+      jdk: oraclejdk8
       after_success: bash <(curl -s https://codecov.io/bash) -f target/coverage/codecov.json
-      jdk: oraclejdk8
-
-    # coveralls
-    - env: TEST_CMD="lein with-profile +1.9,+test-clj,+test-cljs,+cloverage cloverage --coveralls"
-      after_success: curl -F json_file=@target/coverage/coveralls.json https://coveralls.io/api/v1/jobs
-      jdk: oraclejdk8
 
   fast_finish: true      # don't wait for allowed failures before build finish
   allow_failures:
-    - env: TEST_CMD="lein with-profile +1.9,+test-clj,+test-cljs,+eastwood eastwood"
-    - env: TEST_CMD="lein with-profile +1.9,+test-clj,+test-cljs,+cljfmt cljfmt check"
-    - env: TEST_CMD="lein with-profile +1.9,+test-clj,+test-cljs,+cloverage cloverage --codecov"
-    - env: TEST_CMD="lein with-profile +1.9,+test-clj,+test-cljs,+cloverage cloverage --coveralls"
+    - env: VERSION=master TARGET=test-clj
+    - env: VERSION=master TARGET=test-cljs
+    - env: VERSION=1.9 TARGET=cloverage
+    - env: VERSION=1.9 TARGET=eastwood
+    - env: VERSION=1.9 TARGET=cljfmt
 
 notifications:
   webhooks:

--- a/Makefile
+++ b/Makefile
@@ -1,0 +1,40 @@
+.PHONY: clean source-deps test-clj test-cljs eastwood cljfmt cloverage
+
+VERSION ?= 1.9
+export CLOVERAGE_VERSION = 1.0.11-SNAPSHOT
+
+.source-deps:
+	lein source-deps
+	touch .source-deps
+
+source-deps: .source-deps
+
+test-clj: .source-deps
+	lein with-profile +$(VERSION),+plugin.mranderson/config,+test-clj test
+
+test-cljs: .source-deps
+	lein with-profile +$(VERSION),+plugin.mranderson/config,+test-cljs test
+
+eastwood: .source-deps
+	lein with-profile +$(VERSION),+test-clj,+test-cljs,+eastwood eastwood
+
+cljfmt: .source-deps
+	lein with-profile +$(VERSION),+test-clj,+test-cljs,+cljfmt cljfmt check
+
+
+# Cloverage can't handle some of the code in this project.  For now we
+# must filter problematic namespaces (`-e`) and tests (`-t`) from
+# instrumentation. Note: this means for now coverage reporting isn't
+# exact. See issue #457 for details.
+
+cloverage: .source-deps
+	lein with-profile +$(VERSION),+test-clj,+cloverage cloverage --codecov \
+	     -e ".*java.parser" \
+	     -e "cider-nrepl.plugin" \
+	     -e ".*util.instrument" \
+	     -t "^((?!debug-integration-test).)*$$"
+
+clean:
+	lein clean
+	rm .source-deps
+

--- a/project.clj
+++ b/project.clj
@@ -79,7 +79,7 @@
              :test-cljs {:test-paths ["test/cljs"]
                          :dependencies [[com.cemerick/piggieback "0.2.2"]]}
 
-             :cloverage {:plugins [[lein-cloverage "1.0.10"]]}
+             :cloverage {:plugins [[lein-cloverage "1.0.11-SNAPSHOT"]]}
 
              :cljfmt {:plugins [[lein-cljfmt "0.5.7"]]
                       :cljfmt {:indents {as-> [[:inner 0]]

--- a/test/clj/cider/nrepl/middleware/debug_integration_test.clj
+++ b/test/clj/cider/nrepl/middleware/debug_integration_test.clj
@@ -1,5 +1,6 @@
 (ns cider.nrepl.middleware.debug-integration-test
-  (:require [cider.nrepl.middleware.debug :as d]
+  (:require [cider.nrepl :refer [wrap-debug]]
+            [cider.nrepl.middleware.debug :as d]
             [cider.nrepl.test.server :refer [start-server]]
             [clojure.test :refer :all]
             [clojure.tools.nrepl :as nrepl]

